### PR TITLE
Add healthcheck

### PR DESCRIPTION
**Pull request by rafal-gorecki**

### CODE REVIEW

File: healthcheck_manipulation.cpp [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+void healthy_check() {
+  std::chrono::steady_clock::time_point current_time =
+      std::chrono::steady_clock::now();
+  std::chrono::duration<double> odom_elapsed_time =
+      current_time - last_odom_msg_time;
+  std::chrono::duration<double> joints_elapsed_time =
+      current_time - last_odom_msg_time;
+  bool is_odom_valid = odom_elapsed_time.count() < MSG_VALID_TIME.count();
+  bool are_joints_valid = odom_elapsed_time.count() < MSG_VALID_TIME.count();
```
Comment (delihus) [inline healthcheck_manipulation.cpp:57]:
```suggestion
  bool are_joints_valid = joints_elapsed_time.count() < MSG_VALID_TIME.count();
```

File: healthcheck_manipulation.cpp [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+      "joint2",
+      "joint3",
+      "joint4",
+      "imu",
+  };
+
+  for (const auto &joint_name : msg->joint_names) {
+    if (std::find(expected_joint_names.begin(), expected_joint_names.end(),
+                  joint_name) == expected_joint_names.end()) {
+      return;
```
Comment (delihus) [inline healthcheck_manipulation.cpp:42]: What does it happen then the function returns here?
The elapsed time is infinity?

File: healthcheck_manipulation.cpp [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  last_odom_msg_time = std::chrono::steady_clock::now();
+}
+
+void healthy_check() {
+  std::chrono::steady_clock::time_point current_time =
+      std::chrono::steady_clock::now();
+  std::chrono::duration<double> odom_elapsed_time =
+      current_time - last_odom_msg_time;
+  std::chrono::duration<double> joints_elapsed_time =
+      current_time - last_odom_msg_time;
```
Comment (delihus) [inline healthcheck_manipulation.cpp:55]:
```suggestion
      current_time - last_joint_msg_time;
```

**Review by delihus — CHANGES_REQUESTED** [review]

File: Dockerfile.hardware [hunk 4]
```diff
@@ -80,6 +82,18 @@ RUN MYDISTRO=${PREFIX:-ros}; MYDISTRO=${MYDISTRO//-/} && \
     rosdep init && \
     rosdep update --rosdistro $ROS_DISTRO && \
     rosdep install -i --from-path src --rosdistro $ROS_DISTRO -y && \
+    # Create healthcheck package
+    cd src/ && \
+    ros2 pkg create healthcheck_pkg --build-type ament_cmake --dependencies rclcpp std_msgs && \
```
Comment (delihus) [inline Dockerfile.hardware:87]:
```suggestion
    ros2 pkg create healthcheck_pkg --build-type ament_cmake --dependencies rclcpp nav_msgs && \
```

**Review by delihus — CHANGES_REQUESTED** [review]

File: Dockerfile.hardware [hunk 5]
```diff
@@ -80,6 +82,18 @@ RUN MYDISTRO=${PREFIX:-ros}; MYDISTRO=${MYDISTRO//-/} && \
...
+    sed -i '/find_package(std_msgs REQUIRED)/a \
```
Comment (delihus) [inline Dockerfile.hardware:88]: remove std_msgs

File: Dockerfile.hardware [hunk 6]
```diff
@@ -80,6 +82,18 @@ RUN MYDISTRO=${PREFIX:-ros}; MYDISTRO=${MYDISTRO//-/} && \
...
+            find_package(nav_msgs REQUIRED)\n \
+            add_executable(healthcheck_node src/healthcheck.cpp)\n \
+            ament_target_dependencies(healthcheck_node rclcpp std_msgs nav_msgs)\n \
```
Comment (delihus) [inline Dockerfile.hardware:91]:
```suggestion
            ament_target_dependencies(healthcheck_node rclcpp nav_msgs)\n \
```

File: Dockerfile.manipulation.hardware [hunk 7]
```diff
@@ -25,16 +27,39 @@ RUN apt-get update && apt-get install -y \
     rosdep init && \
     rosdep update --rosdistro $ROS_DISTRO && \
     rosdep install -i --from-path src --rosdistro $ROS_DISTRO -y && \
+    # Create healthcheck package
+    cd src/ && \
+    # Does not build
+    rm -rf DynamixelSDK/dynamixel_sdk_examples && \
+    ros2 pkg create healthcheck_pkg --build-type ament_cmake --dependencies rclcpp std_msgs && \
```
Comment (delihus) [inline Dockerfile.manipulation.hardware:34]:
```suggestion
    ros2 pkg create healthcheck_pkg --build-type ament_cmake --dependencies rclcpp control_msgs nav_mgs && \
```

File: Dockerfile.manipulation.hardware [hunk 8]
```diff
@@ -25,16 +27,39 @@ RUN apt-get update && apt-get install -y \
...
+    sed -i '/find_package(std_msgs REQUIRED)/a \
```
Comment (delihus) [inline Dockerfile.manipulation.hardware:35]: remove std_msgs

File: Dockerfile.manipulation.hardware [hunk 9]
```diff
@@ -25,16 +27,39 @@ RUN apt-get update && apt-get install -y \
...
+            find_package(control_msgs REQUIRED)\n \
+            find_package(nav_msgs REQUIRED)\n \
+            add_executable(healthcheck_node src/healthcheck_manipulation.cpp)\n \
+            ament_target_dependencies(healthcheck_node rclcpp std_msgs control_msgs nav_msgs)\n \
```
Comment (delihus) [inline Dockerfile.manipulation.hardware:39]:
```suggestion
            ament_target_dependencies(healthcheck_node rclcpp control_msgs nav_msgs)\n \
```

File: Dockerfile.manipulation.simulation [hunk 10]
```diff
@@ -27,6 +29,21 @@ RUN apt-get update --fix-missing && apt-get install -y \
     rosdep init && \
     rosdep update --rosdistro $ROS_DISTRO && \
     rosdep install -i --from-path src --rosdistro $ROS_DISTRO -y && \
+    # Create healthcheck package
+    cd src/ && \
+    # Does not build
+    rm -rf DynamixelSDK/dynamixel_sdk_examples && \
+    ros2 pkg create healthcheck_pkg --build-type ament_cmake --dependencies rclcpp std_msgs && \
```
Comment (delihus) [inline Dockerfile.manipulation.simulation:36]:
```suggestion
    ros2 pkg create healthcheck_pkg --build-type ament_cmake --dependencies rclcpp control_mgs nav_mgs && \
```

File: Dockerfile.manipulation.simulation [hunk 11]
```diff
@@ -27,6 +29,21 @@ RUN apt-get update --fix-missing && apt-get install -y \
...
+    sed -i '/find_package(std_msgs REQUIRED)/a \
```
Comment (delihus) [inline Dockerfile.manipulation.simulation:37]: remove std_msgs

File: Dockerfile.manipulation.simulation [hunk 12]
```diff
@@ -27,6 +29,21 @@ RUN apt-get update --fix-missing && apt-get install -y \
...
+            find_package(control_msgs REQUIRED)\n \
+            find_package(nav_msgs REQUIRED)\n \
+            add_executable(healthcheck_node src/healthcheck_manipulation.cpp)\n \
+            ament_target_dependencies(healthcheck_node rclcpp std_msgs control_msgs nav_msgs)\n \
```
Comment (delihus) [inline Dockerfile.manipulation.simulation:41]:
```suggestion
            ament_target_dependencies(healthcheck_node rclcpp control_msgs nav_msgs)\n \
```

File: Dockerfile.simulation [hunk 13]
```diff
@@ -33,6 +35,18 @@ RUN MYDISTRO=${PREFIX:-ros}; MYDISTRO=${MYDISTRO//-/} && \
     rosdep init && \
     rosdep update --rosdistro $ROS_DISTRO && \
     rosdep install -i --from-path src --rosdistro $ROS_DISTRO -y && \
+    # Create healthcheck package
+    cd src/ && \
+    ros2 pkg create healthcheck_pkg --build-type ament_cmake --dependencies rclcpp std_msgs && \
```
Comment (delihus) [inline Dockerfile.simulation:40]:
```suggestion
    ros2 pkg create healthcheck_pkg --build-type ament_cmake --dependencies rclcpp nav_msgs && \
```

File: Dockerfile.simulation [hunk 14]
```diff
@@ -33,6 +35,18 @@ RUN MYDISTRO=${PREFIX:-ros}; MYDISTRO=${MYDISTRO//-/} && \
...
+    sed -i '/find_package(std_msgs REQUIRED)/a \
```
Comment (delihus) [inline Dockerfile.simulation:41]: remove std_msgs

File: Dockerfile.simulation [hunk 15]
```diff
@@ -33,6 +35,18 @@ RUN MYDISTRO=${PREFIX:-ros}; MYDISTRO=${MYDISTRO//-/} && \
...
+            find_package(nav_msgs REQUIRED)\n \
+            add_executable(healthcheck_node src/healthcheck.cpp)\n \
+            ament_target_dependencies(healthcheck_node rclcpp std_msgs nav_msgs)\n \
```
Comment (delihus) [inline Dockerfile.simulation:44]:
```suggestion
            ament_target_dependencies(healthcheck_node rclcpp nav_msgs)\n \
```

**Review by delihus — CHANGES_REQUESTED** [review]

File: healthcheck_manipulation.cpp (hunk 2 above)
Comment (rafal-gorecki) [reply healthcheck_manipulation.cpp:42]: then value `last_odom_msg_time` will not be updated, so `healthy_check` detect that last msg is too old and return unhealthy

File: Dockerfile.manipulation.hardware [hunk 16]
```diff
@@ -25,16 +27,37 @@ RUN apt-get update && apt-get install -y \
     rosdep init && \
     rosdep update --rosdistro $ROS_DISTRO && \
     rosdep install -i --from-path src --rosdistro $ROS_DISTRO -y && \
+    # Create healthcheck package
+    cd src/ && \
+    # Does not build
+    rm -rf DynamixelSDK/dynamixel_sdk_examples && \
+    ros2 pkg create healthcheck_pkg --build-type ament_cmake --dependencies rclcpp control_msgs nav_msgs && \
+    sed -i '/find_package(nav_msgs REQUIRED)/a \
```
Comment (delihus) [inline Dockerfile.manipulation.hardware:35]: control_msgs

File: Dockerfile.manipulation.simulation [hunk 17]
```diff
@@ -27,6 +29,19 @@ RUN apt-get update --fix-missing && apt-get install -y \
     rosdep init && \
     rosdep update --rosdistro $ROS_DISTRO && \
     rosdep install -i --from-path src --rosdistro $ROS_DISTRO -y && \
+    # Create healthcheck package
+    cd src/ && \
+    # Does not build
+    rm -rf DynamixelSDK/dynamixel_sdk_examples && \
+    ros2 pkg create healthcheck_pkg --build-type ament_cmake --dependencies rclcpp control_msgs nav_msgs && \
+    sed -i '/find_package(nav_msgs REQUIRED)/a \
```
Comment (delihus) [inline Dockerfile.manipulation.simulation:37]: control_msgs

**Review by delihus — CHANGES_REQUESTED** [review]

**Review by delihus — APPROVED** [review]

**Pull request merged**